File: app/src/main/java/com/example/bt_tester2/MainActivity.java
package com.example.bt_tester2;

import androidx.appcompat.app.AppCompatActivity;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothSocket;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.ListView;

import java.sql.Struct;
import java.util.ArrayList;
import java.util.Set;
import java.util.UUID;

public class MainActivity extends AppCompatActivity {

    private ListView listView;
    private ArrayAdapter aAdapter;
    BluetoothAdapter btAdapter = BluetoothAdapter.getDefaultAdapter();
    BluetoothSocket btSocket;


    ArrayList<String> mac_addresses = new ArrayList<String>();
    ArrayList<String> device_names = new ArrayList<String>();



    static final UUID mUUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB");
    static final String MAC_SAVE = "a1b2";
    static final String DEVICE_SAVE = "a1b1";


    SharedPreferences sharedPref;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        listView = (ListView) findViewById(R.id.listView);

        sharedPref = super.getPreferences(Context.MODE_PRIVATE);

        String savedmac = sharedPref.getString(MAC_SAVE, "");
        String saveddevice = sharedPref.getString(DEVICE_SAVE, "");
        if (!savedmac.equals("")){
            StartUIActivity(saveddevice,savedmac);
        }
        System.out.println(savedmac);


        Set<BluetoothDevice> pairedDevices = btAdapter.getBondedDevices();
        if (pairedDevices.size() > 0) {
            // There are paired devices. Get the name and address of each paired device.
            ArrayList<String> list = new ArrayList<String>();

            for (BluetoothDevice device : pairedDevices) {
                String deviceName = device.getName();
                String deviceHardwareAddress = device.getAddress(); // MAC address
                list.add(deviceName+"\n"+deviceHardwareAddress);
                mac_addresses.add(deviceHardwareAddress);
                device_names.add(deviceName);
            }


            ArrayAdapter<String> adapter = new ArrayAdapter<String>(this,R.layout.customlist, R.id.textviewer1, list);
            listView.setAdapter(adapter);
        }

        listView.setClickable(true);
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {

            public void onItemClick(AdapterView<?> arg0, View arg1, int position, long arg3) {

                String devicename = device_names.get(position).toString();
                String mac = mac_addresses.get(position).toString();

                SharedPreferences.Editor editor = sharedPref.edit();
                editor.putString(MAC_SAVE,mac);
                editor.putString(DEVICE_SAVE,devicename);
                editor.apply();
                System.out.println(devicename + "devicename and " + mac + " mac saved");

                StartUIActivity(devicename,mac);

            }
        });

    }

    void StartUIActivity(String devicename, String mac){
        Intent intent = new Intent(listView.getContext(), UIActivity.class);
        intent.putExtra("device", devicename);
        intent.putExtra("mac",mac);
        startActivity(intent);
    }




}

File: app/src/main/java/com/example/bt_tester2/MainActivity.kt
package com.example.bt_tester2

import android.Manifest
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.View
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.ListView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import com.example.bt_tester2.UIActivity
import java.util.*

open class MainActivity : AppCompatActivity() {
    private var listView: ListView? = null
    //private val aAdapter: ArrayAdapter<*>? = null
    //var btAdapter = BluetoothAdapter.getDefaultAdapter()
    //var btSocket: BluetoothSocket? = null
    var mac_addresses = ArrayList<String>()
    var device_names = ArrayList<String>()
    var sharedPref: SharedPreferences? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        listView = findViewById<View>(R.id.listView) as ListView
        sharedPref = super.getPreferences(MODE_PRIVATE)
        val savedmac = sharedPref?.getString(MAC_SAVE, "")
        var saveddevice = sharedPref?.getString(DEVICE_SAVE, "")
        saveddevice = "ESP32"
        if (savedmac != "") {
            StartUIActivity(saveddevice, savedmac)
        }
        println(savedmac)



        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.BLUETOOTH_CONNECT
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            // TODO: Consider calling
            //    ActivityCompat#requestPermissions
            return
        }
        /*val pairedDevices = btAdapter.bondedDevices
        if (pairedDevices.size > 0) {
            // There are paired devices. Get the name and address of each paired device.
            val list = ArrayList<String>()
            for (device in pairedDevices) {
                val deviceName = device.name
                val deviceHardwareAddress = device.address // MAC address
                list.add(
                    """
                        $deviceName
                        $deviceHardwareAddress
                        """.trimIndent()
                )
                mac_addresses.add(deviceHardwareAddress)
                device_names.add(deviceName)
            }
            val adapter = ArrayAdapter(this, R.layout.customlist, R.id.textviewer1, list)
            listView!!.adapter = adapter
        }*/
        listView!!.isClickable = true
        listView!!.onItemClickListener = OnItemClickListener { arg0, arg1, position, arg3 ->
            val devicename = device_names[position]
            val mac = mac_addresses[position]
            val editor = sharedPref?.edit()
            editor?.putString(MAC_SAVE, mac)
            editor?.putString(DEVICE_SAVE, devicename)
            editor?.apply()
            println(devicename + "devicename and " + mac + " mac saved")
            StartUIActivity(devicename, mac)
        }
    }

    fun StartUIActivity(devicename: String?, mac: String?) {
        val intent = Intent(listView!!.context, UIActivity::class.java)
        intent.putExtra("device", devicename)
        intent.putExtra("mac", mac)
        startActivity(intent)

    }

    companion object {
        val mUUID = UUID.fromString("00001101-0000-1000-8000-00805F9B34FB")
        const val MAC_SAVE = "a1b2"
        const val DEVICE_SAVE = "a1b1"
    }
}